### ISSUE TITLE
Mitochondrial gene query not working with documented arguments

### ISSUE THREAD

**Issue by ivirshup** [scverse/scanpy#242]
I'm having some trouble getting the mitochondrial gene query to not throw an error. Here's an example:

```python
import scanpy.api as sc
sc.queries.mitochondrial_genes(host="www.ensembl.org", org="hsapiens")
```
<details>
<summary>The output and traceback</summary>

```python
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
---------------------------------------------------------------------------
EmptyDataError                            Traceback (most recent call last)
<ipython-input-14-a6967c88fd61> in <module>()
----> 1 sc.queries.mitochondrial_genes(host="www.ensembl.org", org="hsapiens")

/usr/local/lib/python3.6/site-packages/scanpy/queries/__init__.py in mitochondrial_genes(host, org)
     40 
     41     # parsing mitochondrial gene symbols
---> 42     res = pd.read_csv(StringIO(s.query(xml)), sep='\t', header=None)
     43     res.columns = ['symbol', 'chromosome_name']
     44     res = res.dropna()

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in parser_f(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, doublequote, delim_whitespace, low_memory, memory_map, float_precision)
    676                     skip_blank_lines=skip_blank_lines)
    677 
--> 678         return _read(filepath_or_buffer, kwds)
    679 
    680     parser_f.__name__ = name

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in _read(filepath_or_buffer, kwds)
    438 
    439     # Create the parser.
--> 440     parser = TextFileReader(filepath_or_buffer, **kwds)
    441 
    442     if chunksize or iterator:

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in __init__(self, f, engine, **kwds)
    785             self.options['has_index_names'] = kwds['has_index_names']
    786 
--> 787         self._make_engine(self.engine)
    788 
    789     def close(self):

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in _make_engine(self, engine)
   1012     def _make_engine(self, engine='c'):
   1013         if engine == 'c':
-> 1014             self._engine = CParserWrapper(self.f, **self.options)
   1015         else:
   1016             if engine == 'python':

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in __init__(self, src, **kwds)
   1706         kwds['usecols'] = self.usecols
   1707 
-> 1708         self._reader = parsers.TextReader(src, **kwds)
   1709 
   1710         passed_names = self.names is None

pandas/_libs/parsers.pyx in pandas._libs.parsers.TextReader.__cinit__()

EmptyDataError: No columns to parse from file
```

</details>

The arguments I've passed there are whats in the documentation for the function, so I'd figured I'd give them a shot first. I've also tried `host="www.ensembl.org/biomart"`, but had no such luck.

This is after installing the `bioservices` module via `pip` and it creating some config file the first time I tried to run the query.

**Comment by fbrundu** [comment]
I had this error in the past. Try with a different mirror, you can look them from [here](http://www.ensembl.org/info/about/mirrors.html). For example, I usually use `useast.ensembl.org`.

Hope it helps.

**Comment by ivirshup** [comment]
I'd tried `asia.ensembl.org`, but no luck there.

Has the function worked for you?

**Comment by fbrundu** [comment]
Try with `useast.ensembl.org`. I am unable to retrieve anything from `asia`. `useast` works, see below. I think this is a problem on ensembl servers, not of scanpy API.

```python
In [1]: import scanpy.api as sc

In [4]: mito = sc.queries.mitochondrial_genes('useast.ensembl.org', 'mmusculus');

In [5]: mito
Out[5]:
Index(['mt-Tf', 'mt-Rnr1', 'mt-Tv', 'mt-Rnr2', 'mt-Tl1', 'mt-Nd1', 'mt-Ti',
       'mt-Tq', 'mt-Tm', 'mt-Nd2', 'mt-Tw', 'mt-Ta', 'mt-Tn', 'mt-Tc', 'mt-Ty',
       'mt-Co1', 'mt-Ts1', 'mt-Td', 'mt-Co2', 'mt-Tk', 'mt-Atp8', 'mt-Atp6',
       'mt-Co3', 'mt-Tg', 'mt-Nd3', 'mt-Tr', 'mt-Nd4l', 'mt-Nd4', 'mt-Th',
       'mt-Ts2', 'mt-Tl2', 'mt-Nd5', 'mt-Nd6', 'mt-Te', 'mt-Cytb', 'mt-Tt',
       'mt-Tp'],
      dtype='object', name='symbol')

In [6]: mito = sc.queries.mitochondrial_genes('useast.ensembl.org', 'hsapiens');

In [7]: mito
Out[7]:
Index(['MT-TF', 'MT-RNR1', 'MT-TV', 'MT-RNR2', 'MT-TL1', 'MT-ND1', 'MT-TI',
       'MT-TQ', 'MT-TM', 'MT-ND2', 'MT-TW', 'MT-TA', 'MT-TN', 'MT-TC', 'MT-TY',
       'MT-CO1', 'MT-TS1', 'MT-TD', 'MT-CO2', 'MT-TK', 'MT-ATP8', 'MT-ATP6',
       'MT-CO3', 'MT-TG', 'MT-ND3', 'MT-TR', 'MT-ND4L', 'MT-ND4', 'MT-TH',
       'MT-TS2', 'MT-TL2', 'MT-ND5', 'MT-ND6', 'MT-TE', 'MT-CYB', 'MT-TT',
       'MT-TP'],
      dtype='object', name='symbol')

In [8]: mito = sc.queries.mitochondrial_genes('asia.ensembl.org', 'hsapiens');
You must set the host (e.g. f.host='www.ensembl.org'
You must set the host (e.g. f.host='www.ensembl.org'
You must set the host (e.g. f.host='www.ensembl.org'
You must set the host (e.g. f.host='www.ensembl.org'
You must set the host (e.g. f.host='www.ensembl.org'
You must set the host (e.g. f.host='www.ensembl.org'
---------------------------------------------------------------------------
EmptyDataError                            Traceback (most recent call last)
<ipython-input-8-69981c8665eb> in <module>()
----> 1 mito = sc.queries.mitochondrial_genes('asia.ensembl.org', 'hsapiens');

~/.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/site-packages/scanpy/queries/__init__.py in mitochondrial_genes(host, org)
     40
     41     # parsing mitochondrial gene symbols
---> 42     res = pd.read_csv(StringIO(s.query(xml)), sep='\t', header=None)
     43     res.columns = ['symbol', 'chromosome_name']
     44     res = res.dropna()

~/.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/io/parsers.py in parser_f(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, doublequote, delim_whitespace, low_memory, memory_map, float_precision)
    676                     skip_blank_lines=skip_blank_lines)
    677
--> 678         return _read(filepath_or_buffer, kwds)
    679
    680     parser_f.__name__ = name

~/.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/io/parsers.py in _read(filepath_or_buffer, kwds)
    438
    439     # Create the parser.
--> 440     parser = TextFileReader(filepath_or_buffer, **kwds)
    441
    442     if chunksize or iterator:

~/.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/io/parsers.py in __init__(self, f, engine, **kwds)
    785             self.options['has_index_names'] = kwds['has_index_names']
    786
--> 787         self._make_engine(self.engine)
    788
    789     def close(self):

~/.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/io/parsers.py in _make_engine(self, engine)
   1012     def _make_engine(self, engine='c'):
   1013         if engine == 'c':
-> 1014             self._engine = CParserWrapper(self.f, **self.options)
   1015         else:
   1016             if engine == 'python':

~/.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/io/parsers.py in __init__(self, src, **kwds)
   1706         kwds['usecols'] = self.usecols
   1707
-> 1708         self._reader = parsers.TextReader(src, **kwds)
   1709
   1710         passed_names = self.names is None

pandas/_libs/parsers.pyx in pandas._libs.parsers.TextReader.__cinit__()

EmptyDataError: No columns to parse from file
```

**Comment by ivirshup** [comment]
I still get the same error with `useast`

```python
sc.queries.mitochondrial_genes("useast.ensembl.org", "hsapiens")
```
<details>
<summary>The output and traceback</summary>

```python
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
You must set the host (e.g. f.host='www.ensembl.org' 
---------------------------------------------------------------------------
EmptyDataError                            Traceback (most recent call last)
<ipython-input-4-66c3fcd14dab> in <module>()
----> 1 sc.queries.mitochondrial_genes("useast.ensembl.org", "hsapiens")

~/github/scanpy/scanpy/queries/__init__.py in mitochondrial_genes(host, org)
     40 
     41     # parsing mitochondrial gene symbols
---> 42     res = pd.read_csv(StringIO(s.query(xml)), sep='\t', header=None)
     43     res.columns = ['symbol', 'chromosome_name']
     44     res = res.dropna()

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in parser_f(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, doublequote, delim_whitespace, low_memory, memory_map, float_precision)
    676                     skip_blank_lines=skip_blank_lines)
    677 
--> 678         return _read(filepath_or_buffer, kwds)
    679 
    680     parser_f.__name__ = name

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in _read(filepath_or_buffer, kwds)
    438 
    439     # Create the parser.
--> 440     parser = TextFileReader(filepath_or_buffer, **kwds)
    441 
    442     if chunksize or iterator:

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in __init__(self, f, engine, **kwds)
    785             self.options['has_index_names'] = kwds['has_index_names']
    786 
--> 787         self._make_engine(self.engine)
    788 
    789     def close(self):

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in _make_engine(self, engine)
   1012     def _make_engine(self, engine='c'):
   1013         if engine == 'c':
-> 1014             self._engine = CParserWrapper(self.f, **self.options)
   1015         else:
   1016             if engine == 'python':

/usr/local/lib/python3.6/site-packages/pandas/io/parsers.py in __init__(self, src, **kwds)
   1706         kwds['usecols'] = self.usecols
   1707 
-> 1708         self._reader = parsers.TextReader(src, **kwds)
   1709 
   1710         passed_names = self.names is None

pandas/_libs/parsers.pyx in pandas._libs.parsers.TextReader.__cinit__()

EmptyDataError: No columns to parse from file
```

</details>

But it seems to work no matter what server I choose in R, here's the test code

```R
library(biomaRt)

wwwmart    = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl", host="www.ensembl.org")
asiamart   = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl", host="asia.ensembl.org")
useastmart = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl", host="useast.ensembl.org")

wwwresult = getBM(attributes=c("hgnc_symbol", "chromosome_name"), mart=wwwmart)
asiaresult = getBM(attributes=c("hgnc_symbol", "chromosome_name"), mart=asiamart)
useastresult = getBM(attributes=c("hgnc_symbol", "chromosome_name"), mart=useastmart)

assertthat::assert_that(all(wwwresult == asiaresult))
assertthat::assert_that(all(useastresult == asiaresult))
```

I don't think the issue is with ensembl's servers. Is it possible you have a config for this `bioservices` module you've done anything with?

**Comment by ivirshup** [comment]
Just found this:

```python
>>> from bioservices import BioMart
>>> mart = BioMart(host="www.ensembl.org")
>>> mart.registry()
You must set the host (e.g. f.host='www.ensembl.org' 
>>> mart.host = "www.ensembl.org"
>>> mart.registry()
You must set the host (e.g. f.host='www.ensembl.org' 
```

So I'll go open an issue with the `bioservices` package.

**Comment by fbrundu** [comment]
Weird behavior, I have the same version `1.5.2` but no issues..

**Comment by ivirshup** [comment]
I did a little more looking into this, and think I see what's going on. If you set the host, it sends a query to the host to make sure it's real. If it doesn't get a 200 response, it silently fails and does not set the host. Every time I've checked, `www.ensembl.org` returns some `30*` code, which is a redirect, and would successfully fill any query anyways.

`asia.ensembl.org` might have actually been down when I was checking, but it might be worth updating the docs that this probably won't work with `www.ensembl.org`.

Also I'm pretty sure setting the host at instantiation would never work, cause I don't see it happening in their `__init__` function.

**Comment by gokceneraslan** [comment]
`useast.ensembl.org` seems to be working.

**Comment by fbrundu** [comment]
You can check if the mirror is working by using the url:

http://<ensembl_biomart_mirror>/biomart/martview

This is the url used internally by bioservices. The available mirrors are listed [here](http://www.ensembl.org/info/about/mirrors.html).

For example, at the moment the mirror useast is not available.

> Service Temporarily Unavailable
> The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later.

I think we can close this issue since it's not related to the scanpy's code.

**Comment by ivirshup** [comment]
I'd argue a janky dependency is an issue with scanpy's code, as there are interfaces to BioMart which are better behaved. Here's a proof of concept:

```python
def mitochondrial_genes(org, attrname="external_gene_name", host="www.ensembl.org"):
    """Mitochondrial gene symbols for specific organism through BioMart.

    Parameters
    ----------
    org : {{"hsapiens", "mmusculus", "drerio"}}
        Organism to query. Must be an organism in ensembl biomart.
    fieldname : `str`, optional (default: "external_gene_name")
        Biomart attribute field to return. Possible values include 
        "external_gene_name", "ensembl_gene_id", "hgnc_symbol", "mgi_symbol",
        and "zfin_id_symbol".
    host : {{"www.ensembl.org", ...}}
        A valid BioMart host URL.

    Returns
    -------
    An `np.array` containing identifiers for mitochondrial genes.
    """
    try:
        from pybiomart import Server
    except ImportError:
        raise ImportError(
            "You need to install the `pybiomart` module.")
    server = Server(host)
    dataset = (server.marts["ENSEMBL_MART_ENSEMBL"]
                     .datasets["{}_gene_ensembl".format(org)])
    res = dataset.query(
        attributes=[attrname], 
        filters={"chromosome_name": ["MT"]},
        use_attr_names=True
    )
    return res[attrname].values
```

Running it:

```python
>>> mitochondrial_genes("hsapiens")
array(['MT-TF', 'MT-RNR1', 'MT-TV', 'MT-RNR2', 'MT-TL1', 'MT-ND1',
       'MT-TI', 'MT-TQ', 'MT-TM', 'MT-ND2', 'MT-TW', 'MT-TA', 'MT-TN',
       'MT-TC', 'MT-TY', 'MT-CO1', 'MT-TS1', 'MT-TD', 'MT-CO2', 'MT-TK',
       'MT-ATP8', 'MT-ATP6', 'MT-CO3', 'MT-TG', 'MT-ND3', 'MT-TR',
       'MT-ND4L', 'MT-ND4', 'MT-TH', 'MT-TS2', 'MT-TL2', 'MT-ND5',
       'MT-ND6', 'MT-TE', 'MT-CYB', 'MT-TT', 'MT-TP'], dtype=object)
>>> mitochondrial_genes("hsapiens", host="asia.ensembl.org")
array(['MT-TF', 'MT-RNR1', 'MT-TV', 'MT-RNR2', 'MT-TL1', 'MT-ND1',
       'MT-TI', 'MT-TQ', 'MT-TM', 'MT-ND2', 'MT-TW', 'MT-TA', 'MT-TN',
       'MT-TC', 'MT-TY', 'MT-CO1', 'MT-TS1', 'MT-TD', 'MT-CO2', 'MT-TK',
       'MT-ATP8', 'MT-ATP6', 'MT-CO3', 'MT-TG', 'MT-ND3', 'MT-TR',
       'MT-ND4L', 'MT-ND4', 'MT-TH', 'MT-TS2', 'MT-TL2', 'MT-ND5',
       'MT-ND6', 'MT-TE', 'MT-CYB', 'MT-TT', 'MT-TP'], dtype=object)
>>> mitochondrial_genes("hsapiens", host="useast.ensembl.org")
array(['MT-TF', 'MT-RNR1', 'MT-TV', 'MT-RNR2', 'MT-TL1', 'MT-ND1',
       'MT-TI', 'MT-TQ', 'MT-TM', 'MT-ND2', 'MT-TW', 'MT-TA', 'MT-TN',
       'MT-TC', 'MT-TY', 'MT-CO1', 'MT-TS1', 'MT-TD', 'MT-CO2', 'MT-TK',
       'MT-ATP8', 'MT-ATP6', 'MT-CO3', 'MT-TG', 'MT-ND3', 'MT-TR',
       'MT-ND4L', 'MT-ND4', 'MT-TH', 'MT-TS2', 'MT-TL2', 'MT-ND5',
       'MT-ND6', 'MT-TE', 'MT-CYB', 'MT-TT', 'MT-TP'], dtype=object)
```

The drawback with [pybiomart](https://github.com/jrderuiter/pybiomart) is it's unclear how actively it's maintained. However if this looks reasonable to you, I can open a PR with this change.

**Comment by fbrundu** [comment]
I agree but I'm not in charge of the PRs. Probably a better solution would be to provide a backend parameter in the PR to choose which module to use. In my opinion bioservices looks more mature and maintained.

**Comment by falexwolf** [comment]
Hm, my impression is also that `bioservices` is relatively mature but tries to achieve a lot of things.

`pybiomart` looks minimalistic and evidently only focuses on interfacing biomart and has a high coverage, which is nice. For that reason, I tend to prefer @ivirshup implementation over the @fbrundu's original `bioservices`-based one.

But I don't have a strong opinion. I would not have implementations for two "backends" here for such a simple thing (querying biomart).

@fbrundu, would you be alright with changing this to @ivirshup implementation?

**Comment by fbrundu** [comment]
Of course! Whatever works, I don't mind have the implementation changed..

Thanks

**Comment by LuckyMD** [comment]
Maybe I could throw in another ID mapping tool. [BED](https://f1000research.com/articles/7-195/v1) is pretty good. More comprehensive than Biomart and quicker too. It is however a local implementation that runs in a docker container. The image is updated every month or so. At the moment I run a container internally here... but maybe we could make a webserver out of this which can be directly integrated with scanpy?

**Comment by falexwolf** [comment]
Interesting, but hm, I can't really judge on BED... sounds like a more long-term and comprehensive project?

**Comment by LuckyMD** [comment]
It's probably not a quick fix, you are right. Just thought that if future ID mapping is considered here, it would be good to add an option. 

I could add a python API (I have a barebones version already) and we could add it to `sce` for anyone who is running a local BED container anyway?

**Comment by falexwolf** [comment]
Yes, in `sce` this would be welcome!

**Comment by LuckyMD** [comment]
It's on my Todo list!

**Comment by ivirshup** [comment]
Just an update, I've got the PR mostly done, but I'm having trouble keeping the arguments to `sc.queries.gene_coordinates` simple. Could someone who uses that function show me their use case? Is it that common to want the coordinates for only one gene, but also want to limit the coordinates to particular coordinates? 

Would it be reasonable to replace this function with something more simple and open ended? I'm thinking just letting the user specify an organism and the fields they'd like.

<details>
<summary>Here's a doc-string for what I'm thinking:</summary>

```python
def biomart_annotations(org, attrs, host="www.ensembl.org"):
    """
    Retrieve gene annotations from ensembl biomart.

    Parameters
    ----------
    org : `str`
        Organism to query. Must be an organism in ensembl biomart. "hsapiens",
        "mmusculus", "drerio", etc.
    attrs : `List[str]`
        Attributes to query biomart for.
    host : `str`, optional (default: "www.ensembl.org")
        A valid BioMart host URL. Alternative values include archive urls (like
        "grch37.ensembl.org") or regional mirrors (like "useast.ensembl.org").

    Returns
    -------
    A `pd.DataFrame` containing annotations.

    Examples
    --------
    Retrieve genes coordinates and chromosomes

    >>> annot = sc.query.biomart_annotations(
            "hsapiens",
            ["ensembl_gene_id", "start_position", "end_position", "chromosome_name"],
        ).set_index("ensembl_gene_id")
    >>> adata.var[annot.columns] = annot
```

</details>

**Comment by ivirshup** [comment]
@LuckyMD that BED tool does look cool, and I can think of a couple times it'd would've been really useful. Is it easy to get information beyond gene-ids out of it? For instance, could I give a uniprot id and get transcript locations from ensembl?

**Comment by flying-sheep** [comment]
Yes. The only problem with BED is that it’s super large and needs quite some memory to run as well. Having a publicly hosted instance backed by some rich company would be great.

**Comment by LuckyMD** [comment]
The latest BED docker image is 1.62 GB, which is not that big for an ID mapping DB I reckon (with the docker environment of course). You can also build your own, more limited versions of the database. There are instructions for this in the [github](https://github.com/patzaw/bed).

BED memory usage I haven't really checked. Server side it's using 1 GB idly at the moment, no idea how that changes at peak usage. It has a client-side cache system to speed up multiple look-ups, but that can be turned off to limit client-side memory usage.

And yes, as @flying-sheep said it's really easy to use for mapping between biological entities and even organisms as well. For transcript locations, you may need to use a separate file that contains the locations which will probably be annotated with RefSeq IDs. You can get to the RefSeq IDs easily though.

It was actually designed for company use... but I think Thompson-Reuters (now Clarivate analytics) or UCB aren't interested in hosting public servers ;).

**Comment by flying-sheep** [comment]
This was fixed in #467, right?

**Comment by ivirshup** [comment]
👍
